Add debug_traceBlockByNumber archive check for MetaMask compliance

natefikru opened the pull request:
## Summary

Adds an optional `debug_traceBlockByNumber` check to the archive health check flow for MetaMask compliance requirements. This verifies that providers support trace/debug capabilities in addition to the existing `eth_getBalance` archive check.

### New Configuration

```caddyfile
networks {
    eth {
        archive_enabled true
        archive_trace_block_by_number true
    }
}
```

### Behavior

| archive_enabled | archive_trace_block_by_number | Result |
|-----------------|-------------------------------|--------|
| false | false | No archive checks |
| false | true | No archive checks |
| true | false | Only eth_getBalance |
| true | true | eth_getBalance + debug_traceBlockByNumber |

### Request Payload

```json
{
  "jsonrpc": "2.0",
  "method": "debug_traceBlockByNumber",
  "id": 1,
  "params": ["0xBLOCK_HEX", {"tracer": "callTracer", "timeout": "30s", "onlyTopCall": true}]
}
```

### Failure Handling

- Uses same block number as existing archive check (quarter-height: `currentBlock / 4`)
- Failures return **Warning** status (not Unhealthy), keeping provider in traffic rotation
- Properly detects JSON-RPC errors (e.g., `-32000 block pruned`) even on HTTP 200 responses
- Failures logged via `logProviderWarning()` with `health_status=Warning` label

### Files Changed

- `modules/consts.go` - Added default constant
- `modules/network.go` - Added struct fields, flag tracking, trace check integration
- `modules/caddy_unmarshaller.go` - Added Caddyfile parsing
- `lib/network/json_rpc_client.go` - Added shared helper function
- `lib/network/evm_handler.go` - Added EVM-specific trace methods

natefikru committed:
enable arhive debug check and flag

natefikru committed:
added additional tests

noleto commented on lib/network/evm_handler.go:
```diff
@@ -468,6 +468,42 @@ func (h *EVMHandler) PerformArchiveCheck(httpUrl string, headers map[string]stri
 
+// PerformTraceBlockByNumberCheck performs debug_traceBlockByNumber check for EVM chains
+// This is an additional archive mode check that verifies trace capabilities (MetaMask requirement)
+func (h *EVMHandler) PerformTraceBlockByNumberCheck(httpUrl string, headers map[string]string, httpClient din_http.IHTTPClient, authClient auth.IAuthClient, requestAttempts int, blockHeight string) error {
```
I think it makes sense to add this method to the `NetworkHandler interface` contract so that every handler can decide whether it is eligible to implement it or not

natefikru replied:
I would push against this because this is more of an evm specific call, it is a sub function of the archive check function which has an interface associated with it.

smoyer64 replied:
I think we should abstract the idea of healthcheck even more - I'll have a proposal that shrinks the `Handler` interface in January

noleto replied:
Okay, it's fair if that's specific to EVM. It should at least have its own interface defined in the `handlers.go` file, as you currently have it declared inline here: https://github.com/DIN-center/din-caddy-plugins/blob/0aab16f490f91699c5d88fd2dce29d8e0ffe5883/modules/network.go#L459-L461) This is not to be picky: it makes the implementation discoverable by IDE tools and also makes it easier to mock the behavior in tests. Defining an inline interface, while possible, is not very idiomatic in Go code

noleto commented on lib/network/evm_handler.go:
```diff
@@ -468,6 +468,42 @@ func (h *EVMHandler) PerformArchiveCheck(httpUrl string, headers map[string]stri
+
+	return nil
+}
+
```
These two methods seem to be just helper functions, so we don't need to add them as handler methods. This will help keep the EVM handler more focused on implementing only the methods in the contract. Additionally, it would be great to move them to a helper file.

natefikru replied:
good call, done.

noleto commented on lib/network/json_rpc_client.go:
```diff
@@ -179,6 +179,55 @@ func PerformArchiveCheckViaJSONRPC(httpUrl string, headers map[string]string, ht
 
+// PerformTraceBlockByNumberCheckViaJSONRPC performs a debug_traceBlockByNumber check via JSON-RPC
+// This is EVM-specific and verifies trace/debug capabilities for MetaMask compliance
+func PerformTraceBlockByNumberCheckViaJSONRPC(httpUrl string, headers map[string]string, httpClient din_http.IHTTPClient, authClient auth.IAuthClient, requestAttempts int, blockHeight string, createPayloadFunc func(string) ([]byte, error), parseResponseFunc func([]byte) error) error {
```
What is the rationale behind passing the closures `createPayloadFunc` and `parseResponseFunc` to this function? Does it vary depending on the caller? Currently, there is only a single call made passing `CreateTraceBlockByNumberPayload` and `ParseTraceBlockByNumberResponse`, respectively. If the behavior is likely to change depending on the caller, it might be better to introduce an interface to capture both of these behaviors you want to customize

noleto replied:
see https://refactoring.guru/design-patterns/template-method/go/example

natefikru replied:
you're right, thats definitely over engineered, done

noleto commented on modules/network.go:
```diff
@@ -444,7 +447,26 @@ func (n *network) performArchiveCheck(provider *provider, currentBlock int64) er
+		// Check if handler supports trace check via type assertion (EVM-specific)
+		type traceChecker interface {
+			PerformTraceBlockByNumberCheck(httpUrl string, headers map[string]string, httpClient din_http.IHTTPClient, authClient auth.IAuthClient, requestAttempts int, blockHeight string) error
+		}
```
I'd suggest moving the interface declaration to `handlers.go` as an optional interface to implement. Alternatively, and this is my preference, include it in the `NetworkHandler`, which will make the method discoverable from the `NetworkHandler` interface and eliminate the need for type assertions, resulting in more idiomatic Go code. In both cases, it will be easier to mock and test that behavior.

natefikru committed:
refactor: convert trace block helpers to private standalone functions

- Remove struct receiver from createTraceBlockByNumberPayload and
  parseTraceBlockByNumberResponse, making them private helper functions
- Remove closure parameters from PerformTraceBlockByNumberCheckViaJSONRPC
  and call helpers directly (simplifies code since behavior doesn't vary)
- Update tests to call helpers directly

natefikru committed:
Merge remote-tracking branch 'origin/develop' into featre/add_additional_archive_checks

natefikru committed:
chore: ignore plans, docs, and router.log files

natefikru committed:
refactor: move TraceChecker interface to handlers.go

Move TraceChecker from inline interface definition in modules/network.go
to exported interface in lib/network/handlers.go for better discoverability
and following Go idioms for optional interface patterns.

Changes:
- Add TraceChecker interface to lib/network/handlers.go
- Update modules/network.go to use networklib.TraceChecker
- No behavior changes, all tests pass

Addresses code review feedback to improve interface organization.

noleto reviewed (approved):
looks good to me, thanks for addressing my feedback.
